### PR TITLE
Ed25519 sign support

### PR DESCRIPTION
**Pull request by jnippula**
Added implementation for PX4Crypto::sign() and PX4Crypto::generate_keypair()
Added key_type (PRIVATE/PUBLIC) into key_cache and modified der_parse function list to support parser for both.

tests/test_crypto: Implemented ed25519 key generate and signing/signatyre_verification tests

### CODE REVIEW

**Comment by jnippula** [comment]
waiting for pfsoc_crypto changes to be merged

File: src/drivers/sw_crypto/crypto.c [hunk 1]
```diff
@@ -214,6 +214,44 @@ void crypto_close(crypto_session_handle_t *handle)
 	handle->context = NULL;
 }
 
+bool crypto_signature_gen(crypto_session_handle_t handle,
```
[the remark this answers is not in the document]
Comment (jnippula) [reply src/drivers/sw_crypto/crypto.c:217]: I used that crypto_sign in first place, but it conflicted with monocypher API, so I end up changing it. But need to re-check how to handle that.

[the remark this answers is not in the document]
Comment (jlaitine) [reply src/drivers/sw_crypto/crypto.c:217]: Ah ok, then just leave it as is!

File: platforms/common/include/px4_platform_common/crypto_backend.h [hunk 2]
```diff
@@ -327,6 +369,16 @@ typedef struct cryptoiocdecryptdata {
 	size_t ret;
 } cryptoiocdecryptdata_t;
 
+#define CRYPTOIOCGENKEYPAIR _CRYPTOIOC(12)
+typedef struct cryptoiocgenkeypair {
+	crypto_session_handle_t *handle;
+	size_t key_size;
+	uint8_t private_idx;
```
Comment (jlaitine) [inline platforms/common/include/px4_platform_common/crypto_backend.h:376]: What was the reason again not to store private and public parts together with the same index?

Comment (jnippula) [reply platforms/common/include/px4_platform_common/crypto_backend.h:376]: It is using only one key index (the private one) underneath, but I didn't bother to touch the upstreamed API. But, I can change that also and remove the other idx

Comment (jnippula) [reply platforms/common/include/px4_platform_common/crypto_backend.h:376]: well.. the sw_crypto impl I made does not use one slot.. need to fix that.

**Comment by jnippula** [comment]
This contains crypto API change, so need to merge together with pfsoc_crypto changes. Waiting for pfsoc_crypto change to be merged first

File: src/drivers/sw_crypto/crypto.c [hunk 3]
```diff
@@ -251,6 +252,17 @@ bool crypto_signature_check(crypto_session_handle_t handle,
 	return ret;
 }
 
+bool crypto_signature_check(crypto_session_handle_t handle,
+			    uint8_t  key_index,
+			    const uint8_t  *signature,
+			    const uint8_t *message,
+			    size_t message_size)
+{
+
```
Comment (jlaitine) [inline src/drivers/sw_crypto/crypto.c:261]: Isn't this function already impelmented in sw_crypto?

File: platforms/nuttx/src/px4/common/px4_usr_crypto.cpp [hunk 4]
```diff
@@ -87,6 +87,16 @@ bool PX4Crypto::signature_check(uint8_t  key_index,
 	return data.ret;
 }
 
+bool PX4Crypto::sign(uint8_t  key_index,
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/common/px4_usr_crypto.cpp:90]: extra space before key_index

File: platforms/nuttx/src/px4/common/px4_crypto.cpp [hunk 5]
```diff
@@ -117,6 +117,14 @@ void PX4Crypto::close()
 	unlock();
 }
 
+bool PX4Crypto::sign(uint8_t  key_index,
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/common/px4_crypto.cpp:120]: extra space before key_index

File: platforms/common/include/px4_platform_common/crypto_backend.h [hunk 6]
```diff
@@ -304,7 +334,17 @@ typedef struct cryptoiocrenewnonce {
 	size_t ret;
 } cryptoiocrenewnonce_t;
 
-#define CRYPTOIOCSIGNATURECHECK _CRYPTOIOC(9)
+#define CRYPTOIOCSIGN _CRYPTOIOC(9)
+typedef struct cryptoiocsign {
+	crypto_session_handle_t *handle;
+	uint8_t  key_index;
```
Comment (jlaitine) [inline platforms/common/include/px4_platform_common/crypto_backend.h:340]: extra space before argument on both key_index and *signature

File: platforms/common/include/px4_platform_common/crypto_backend.h [hunk 7]
```diff
@@ -181,17 +194,34 @@ bool crypto_get_nonce(crypto_session_handle_t handle,
...
+ * Perform signing using an open session to crypto
+ * handle: session handle, returned by open
+ * key_index: index to the key used for signing
+ * signature: pointer to output signature data
+ * message: pointer to the data to be signed
+ * message_size: size of the data
+ */
+
+bool crypto_signature_gen(crypto_session_handle_t handle,
+			  uint8_t  key_index,
```
Comment (jlaitine) [inline platforms/common/include/px4_platform_common/crypto_backend.h:208]: extra space before key_index

File: platforms/common/include/px4_platform_common/crypto.h [hunk 8]
```diff
@@ -173,6 +170,20 @@ class PX4Crypto
...
+	/*
+	 * Message signing
+	 *
+	 * key_index: private key index in keystore
+	 * signature: pointer to the output signature
+	 * message: pointer to the data to be signed
+	 * message_size: size of the message in bytes
+	*/
+
+	bool sign(uint8_t  key_index,
```
Comment (jlaitine) [inline platforms/common/include/px4_platform_common/crypto.h:182]: extra space before key_index

**Pull request merged**